File: 100-Days-Series/Day 52/DoublyLinkedList.java
import java.util.Scanner;
class DoublyLL {
    int data;
    DoublyLL prev, next;
}
public class DoublyLinkedList {
    private static DoublyLL start = null, end = null;
    private static final Scanner scanner = new Scanner(System.in);
    
    // Function to add new nodes
    public static void addNode() {
        DoublyLL ptr = new DoublyLL();
        System.out.print("\nEnter data: ");
        ptr.data = scanner.nextInt();
        if (start == null) {
            start = end = ptr;
            start.prev = end;
            end.next = start;
        } else {
            ptr.prev = end;
            ptr.next = start;
            end.next = ptr;
            start.prev = ptr;
            end = ptr;
        }
        System.out.println("\nNew Node Added");
    }

    // Function to traverse or display from start to end nodes
    public static void traverseHeadToEnd() {
        if (start == null) {
            System.out.println("\nDoubly Linked List is empty!");
        } else {
            DoublyLL temp = start;
            do {
                System.out.println("\nData: " + temp.data);
                System.out.println("-----------------------");
                temp = temp.next;
            } while (temp != start);
        }
    }

    // Function to traverse or display from end to start nodes
    public static void traverseEndToHead() {
        if (start == null) {
            System.out.println("\nDoubly Linked List is empty!");
        } else {
            DoublyLL temp = end;
            do {
                System.out.println("\nData: " + temp.data);
                System.out.println("----------------------");
                temp = temp.prev;
            } while (temp != end);
        }
    }

    // Function to delete a node
    public static void delNode() {
        if (start == null) {
            System.out.println("\nDoubly Linked List is empty!");
        } else {
            System.out.print("\nEnter data to delete: ");
            int no = scanner.nextInt();
            DoublyLL temp = start;
            boolean found = false;

            do {
                if (temp.data == no) {
                    found = true;
                    if (temp == start && temp == end) {
                        // Only one node in the list
                        start = end = null;
                    } else if (temp == start) {
                        start = start.next;
                        start.prev = end;
                        end.next = start;
                    } else if (temp == end) {
                        end = end.prev;
                        end.next = start;
                        start.prev = end;
                    } else {
                        temp.prev.next = temp.next;
                        temp.next.prev = temp.prev;
                    }
                    System.out.println("\nNode Deleted!");
                    break;
                }
                temp = temp.next;
            } while (temp != start);

            if (!found) {
                System.out.println("\nData not found!");
            }
        }
    }

    // Main Method
    public static void main(String[] args) {
        int ch;
        do {
            System.out.println("\n1. Add New Node");
            System.out.println("2. Traverse (Head to End)");
            System.out.println("3. Traverse (End to Head)");
            System.out.println("4. Delete Node");
            System.out.println("5. Exit");
            System.out.print("Select a choice: ");
            ch = scanner.nextInt();
            switch (ch) {
                case 1:
                    addNode();
                    break;
                case 2:
                    traverseHeadToEnd();
                    break;
                case 3:
                    traverseEndToHead();
                    break;
                case 4:
                    delNode();
                    break;
                case 5:
                    System.out.println("\nExited");
                    break;
                default:
                    System.out.println("\nInvalid choice");
            }
        } while (ch != 5);
    }
}

File: 100-Days-Series/Day 52/DoublyLinkedList.c
#include<stdio.h>
#include<stdlib.h>
struct DoublyLL{
    int data;
    struct DoublyLL *prev,*next;
}*start,*end;
// Function for add new nodes
void addNode(){
    struct DoublyLL *ptr;
    ptr=(struct DoublyLL*)malloc(sizeof(struct DoublyLL));
    if(ptr==NULL){
        printf("\nMemory couldn't Allocated");
    }else{
        printf("\nEnter data : ");
        scanf("%d",&ptr->data);
        if(start==NULL){
            start=end=ptr;
            start->prev=ptr;
            end->next=ptr;
        }else{
            ptr->prev=end;
            ptr->next=start;
            end->next=ptr;
            end=ptr;
            start->prev=end;
        }
        printf("\nNew Node Added");
    }
}
// Function for traverse or display from start to end nodes.
void traverseHeadToEnd(){
    if(start==NULL){
        printf("\nDoubly Linked List is empty!");
    }else{
        struct DoublyLL *temp=start;
        do{
            printf("\nData is %d",temp->data);
            printf("\n-----------------------");
            temp=temp->next;
        }while(temp!=start);
    }
}
// Function for traverse or display from end to tart nodes.
void traverseEndToHead(){
    if(start==NULL){
        printf("\nDoubly Linked List is empty!");
    }else{
        struct DoublyLL *temp=end;
        do{
            printf("\nData is %d",temp->data);
            printf("\n----------------------");
            temp=temp->prev;
        }while(temp!=end);
    }
}
// Function for delete a node
void delNode(){
    if(start==NULL){
        printf("\nDoubly Linked List is empty!");
    }else{
        int no,res=0;
        struct DoublyLL *temp=start;
        printf("\nEnter data to delete : ");
        scanf("%d",&no);
        do{
            if(temp->data==no){
                if(temp==start){
                    if(temp->next==start){
                        start=end=NULL;
                    }else{
                        start=start->next;
                        end->next=start;
                    }
                }else if(temp==end){
                    end=end->prev;
                    end->next=start;
                    start->prev=end;
                }else{
                    temp->next->prev=temp->prev;
                    temp->prev->next=temp->next;
                }
                printf("\nDeleted!");
                res=1;
                free(temp);
                break;
            }
            temp=temp->next;
        }while(temp!=start);
        if(res==0){
            printf("\nData is not found!");
        }
    }
}
// Let' Start Here
void main(){
    start=end=NULL;
    int ch;
    do{
        printf("\n1. Add New Node");
        printf("\n2. Traverse (Head to End)");
        printf("\n3. Traverse (End to Head)");
        printf("\n4. Delete");
        printf("\n5. Exit");
        printf("\nSelect a choice : ");
        scanf("%d",&ch);
        if(ch==1){
            addNode();
        }else if(ch==2){
            traverseHeadToEnd();
        }else if(ch==3){
            traverseEndToHead();
        }else if(ch==4){
            delNode();
        }else if(ch==5){
            printf("\nExited");
        }else{
            printf("\nInvalid choice");
        }
    }while(ch!=5);
}